### PR TITLE
Add more descriptive messages for assertInstantiateError and assertInstantiateSuccess tests

### PR DESCRIPTION
**Pull request by eholk**
This PR gives each of these tests a unique message (although they are a bit verbose, so I'm open to changing them), to make it easier to see exactly what is failing.

@bnjbvr, @rossberg-chromium - could you take a look?

### CODE REVIEW

**Comment by bnjbvr** [comment]
I appreciate that these messages are just plain different than what we had before, but given that they might appear on the web site at some point, I'd go even further and have them document for humans what they test. For instance:

```
assertInstantiateError([], TypeError, "instantiate with no args throws a TypeError"); 
```

What do you think?

**Comment by eholk** [comment]
@bnjbvr Thanks, I think your suggestion makes for much more readable names. I pushed some new updates to the name.

File: test/js-api/jsapi.js [hunk 1]
```diff
@@ -728,27 +728,27 @@ test(() => {
...
-    assertInstantiateError([complexImportingModuleBinary, {"c": {"d": scratch_memory}}], TypeError, "assertInstantiateError([complexImportingModuleBinary, {'c': {'d': scratch_memory}}] => TypeError)");
+    assertInstantiateError([], TypeError, "instantiate with no args throws TypeError)");
+    assertInstantiateError([undefined], TypeError, "instantiate on undefined throws TypeError)");
+    assertInstantiateError([1], TypeError, "instantiate on a number throws TypeError)");
+    assertInstantiateError([{}], TypeError, "instantiate on empty object throws TypeError)");
+    assertInstantiateError([new Uint8Array()], CompileError, "instantiate on empty Uint8Array throws CompileError)");
+    assertInstantiateError([new ArrayBuffer()], CompileError, "instantiate on empty ArrayBuffer throws CompileError)");
+    assertInstantiateError([new Uint8Array("hi!")], CompileError, "instantiate on Uint8Array('hi!') throws CompileError)");
+    assertInstantiateError([new ArrayBuffer("hi!")], CompileError, "instantiate on ArrayBuffer('hi!') throws CompileError)");
```
Comment (shurcooL) [inline test/js-api/jsapi.js]: There's an extra closing parenthesis inside the strings above (but not the ones below). Is that unintentional?

Comment (eholk) [inline test/js-api/jsapi.js]: Nope, just sloppy search and replace on my part. It should be fixed now. Thanks for catching this!

File: test/js-api/jsapi.js [hunk 2]
```diff
@@ -724,33 +724,33 @@ test(() => {
...
+    assertInstantiateError([{}], TypeError, "instantiate on empty object throws TypeError");
+    assertInstantiateError([new Uint8Array()], CompileError, "instantiate on empty Uint8Array throws CompileError");
+    assertInstantiateError([new ArrayBuffer()], CompileError, "instantiate on empty ArrayBuffer throws CompileError");
+    assertInstantiateError([new Uint8Array("hi!")], CompileError, "instantiate on Uint8Array('hi!') throws CompileError");
+    assertInstantiateError([new ArrayBuffer("hi!")], CompileError, "instantiate on ArrayBuffer('hi!') throws CompileError");
+    assertInstantiateError([importingModule], TypeError, "instantiate on importingModule throws TypeError");
+    assertInstantiateError([importingModule, null], TypeError, "instantiate on importingModule, null throws TypeError");
+    assertInstantiateError([importingModuleBinary, null], TypeError, "instantiate on importingModuleBinary, null throws TypeError");
+    assertInstantiateError([emptyModule, null], TypeError, "instantiate on emptyModule throws TypeError");
+    assertInstantiateError([importingModuleBinary, null], TypeError, "instantiate on importingModuleBinary, null throws TypeError");
```
Comment (bnjbvr) [inline test/js-api/jsapi.js]: pre-existing, but this test is actually duplicated above and could be removed.

File: test/js-api/jsapi.js [hunk 3]
```diff
@@ -724,33 +724,33 @@ test(() => {
...
-    function assertInstantiateSuccess(module, imports) {
+    assertInstantiateError([], TypeError, "instantiate with no args throws TypeError");
+    assertInstantiateError([undefined], TypeError, "instantiate on undefined throws TypeError");
+    assertInstantiateError([1], TypeError, "instantiate on a number throws TypeError");
+    assertInstantiateError([{}], TypeError, "instantiate on empty object throws TypeError");
+    assertInstantiateError([new Uint8Array()], CompileError, "instantiate on empty Uint8Array throws CompileError");
+    assertInstantiateError([new ArrayBuffer()], CompileError, "instantiate on empty ArrayBuffer throws CompileError");
+    assertInstantiateError([new Uint8Array("hi!")], CompileError, "instantiate on Uint8Array('hi!') throws CompileError");
+    assertInstantiateError([new ArrayBuffer("hi!")], CompileError, "instantiate on ArrayBuffer('hi!') throws CompileError");
+    assertInstantiateError([importingModule], TypeError, "instantiate on importingModule throws TypeError");
```
Comment (bnjbvr) [inline test/js-api/jsapi.js]: How about "instantiate a module importing functions without passing an imports object throws TypeError"?

File: test/js-api/jsapi.js [hunk 4]
```diff
@@ -724,33 +724,33 @@ test(() => {
...
+    assertInstantiateError([importingModule, null], TypeError, "instantiate on importingModule, null throws TypeError");
+    assertInstantiateError([importingModuleBinary, null], TypeError, "instantiate on importingModuleBinary, null throws TypeError");
+    assertInstantiateError([emptyModule, null], TypeError, "instantiate on emptyModule throws TypeError");
+    assertInstantiateError([importingModuleBinary, null], TypeError, "instantiate on importingModuleBinary, null throws TypeError");
+    assertInstantiateError([importingModuleBinary, undefined], TypeError, "instantiate on importingModuleBinary, undefined throws TypeError");
+    assertInstantiateError([importingModuleBinary, {}], TypeError, "instantiate on importingModuleBinary, {} throws TypeError");
+    assertInstantiateError([importingModuleBinary, {"":{g:()=>{}}}], LinkError, "instantiate on importingModuleBinary, {'':{g:()=>{}}} throws LinkError");
```
Comment (bnjbvr) [inline test/js-api/jsapi.js]: How about "instantiate on an module importing a function named f, not provided in the imports object, throws LinkError"? (etc for other tests below)

File: test/js-api/jsapi.js [hunk 5]
```diff
@@ -724,33 +724,33 @@ test(() => {
...
+    assertInstantiateError([new Uint8Array("hi!")], CompileError, "instantiate on Uint8Array('hi!') throws CompileError");
+    assertInstantiateError([new ArrayBuffer("hi!")], CompileError, "instantiate on ArrayBuffer('hi!') throws CompileError");
+    assertInstantiateError([importingModule], TypeError, "instantiate on importingModule throws TypeError");
+    assertInstantiateError([importingModule, null], TypeError, "instantiate on importingModule, null throws TypeError");
+    assertInstantiateError([importingModuleBinary, null], TypeError, "instantiate on importingModuleBinary, null throws TypeError");
+    assertInstantiateError([emptyModule, null], TypeError, "instantiate on emptyModule throws TypeError");
+    assertInstantiateError([importingModuleBinary, null], TypeError, "instantiate on importingModuleBinary, null throws TypeError");
+    assertInstantiateError([importingModuleBinary, undefined], TypeError, "instantiate on importingModuleBinary, undefined throws TypeError");
+    assertInstantiateError([importingModuleBinary, {}], TypeError, "instantiate on importingModuleBinary, {} throws TypeError");
```
Comment (bnjbvr) [inline test/js-api/jsapi.js]: How about "Providing an empty imports object to a module that expects imports throws TypeError"?

File: test/js-api/jsapi.js [hunk 6]
```diff
@@ -768,19 +768,19 @@ test(() => {
...
+    assertInstantiateSuccess(emptyModuleBinary.buffer, undefined, "instantiate on emptyModuleBinary.buffer, undefined succeeds");
+    assertInstantiateSuccess(importingModule, {"":{f:()=>{}}}, 'instantiate on importingModule, {"":{f:()=>{}}} succeeds');
+    assertInstantiateSuccess(importingModuleBinary, {"":{f:()=>{}}}, 'instantiate on importingModuleBinary, {"":{f:()=>{}}} succeeds');
+    assertInstantiateSuccess(importingModuleBinary.buffer, {"":{f:()=>{}}}, 'instantiate on importingModuleBinary.buffer, {"":{f:()=>{}}} succeeds');
     assertInstantiateSuccess(complexImportingModuleBinary, {
         a:{b:()=>{}},
         c:{d:scratch_memory},
         e:{f:scratch_table},
-        g:{'⚡':1}});
+        g:{'⚡':1}}, "instantiate on complexImportingModuleBinary, {...} succeeds");
```
Comment (bnjbvr) [inline test/js-api/jsapi.js]: How about "instantiate complexImportingModuleBinary with all expected imports succeeds"?

File: test/js-api/jsapi.js (hunk 3 above)
Comment (eholk) [inline test/js-api/jsapi.js]: Done.

File: test/js-api/jsapi.js (hunk 2 above)
Comment (eholk) [inline test/js-api/jsapi.js]: Done.

File: test/js-api/jsapi.js (hunk 5 above)
Comment (eholk) [inline test/js-api/jsapi.js]: Done.

File: test/js-api/jsapi.js (hunk 4 above)
Comment (eholk) [inline test/js-api/jsapi.js]: Done (I got a few more of the tests below, but may have missed some)

File: test/js-api/jsapi.js (hunk 6 above)
Comment (eholk) [inline test/js-api/jsapi.js]: Done.

**Comment by eholk** [comment]
I just pushed a few more changes, tweaking the names some more. I also added the `Set` to make sure the test names are unique, at least when run from the console.

Do we have an ideal line length for this repo? It looks like the existing code was already beyond 80 columns. This change now has a 206 character line, which is pretty unreasonable :)

File: test/js-api/jsapi.js [hunk 7]
```diff
@@ -714,7 +714,22 @@ test(() => {
...
     assert_equals(instantiate.length, 1);
     assert_equals(instantiate.name, "instantiate");
-    function assertInstantiateError(args, err) {
+
+    // Keep track of all the messages for assertInstantiateError and assertCompileSuccess
+    // to make sure they're all unique.
+    const testNames = new Set();
+    function ensureUniqueTest(message) {
+        if (testNames.has(message)) {
+            if (quit instanceof Function) {
```
Comment (bnjbvr) [inline test/js-api/jsapi.js]: The current code will throw a TypeError if `quit` isn't defined (yay JS!). How about `typeof quit === 'function'` instead?

**Comment by rossberg-chromium** [comment]
I like more informative messages, but I wonder how this approach can scale. The large chunk of the tests that the browser will run are from the core suite, and that contains 13 000+ individual assertions. It doesn't seem realistic to annotate all of them with informative (let alone unique) messages.

**Comment by bnjbvr** [comment]
Indeed, it's not clear for the core tests. For these, a unique sequential number is generated per assertion, at the moment, which is not too helpful to humans, but it allows the WPT harness to automatically intercept errors and generate expectation files.

Maybe a good rule of thumb is to add new descriptive messages at least for new test cases? Usually, a single `test()` (or `promise_test()`) can group several actions and assertions. When seeing a failure in the WPT HTML report, one can see the unique error message, grep for it and find which exact line is affected in the test. (It's possible to hide failures or successes in the WPT HTML report)

**Comment by rossberg-chromium** [comment]
How about we simply report file name/line numbers from the original .wast? The spec interpreter already embeds this information in the generated JS, but currently only as a comment. It would be easy to change this so that it becomes an additional string argument to the harness functions.

Not sure a more heavy-weight approach is worthwhile. Implementers will need to look at the source anyway. And for other random users running the tests I doubt that any form of clear text will be particularly useful.

**Comment by bnjbvr** [comment]
> How about we simply report file name/line numbers from the original .wast?

Would be great indeed.

For the JS tests, we could even get the stack trace (with a caught Error) and generate a description containing the caller's source line as well.

File: test/js-api/jsapi.js (hunk 7 above)
Comment (eholk) [inline test/js-api/jsapi.js]: Done. I wasn't sure what the idiomatic way to check whether I should or can call something is.

**Comment by eholk** [comment]
I just reflowed this to 100 columns, since that seems to match the rest of the file.

@rossberg-chromium I like the idea of using filename and line numbers from the original .wast, but that approach isn't really relevant here because these tests are for the JS API, which aren't generated from a .wast. Also, I'd be worried that somebody could insert a line in one of the .wasts and then change all the test IDs. 

Ultimately, I'd like something unique and stable to identify tests (#415) so that VMs can list the ones they currently fail on (for example, [in V8](https://cs.chromium.org/chromium/src/v8/test/mjsunit/wasm/jsapi-harness.js?q=jsapi-harness.js+package:%5Echromium$&l=7)). I've been using the human readable description, since it's the best we have so far, but it's not the right solution for the long term.

**Comment by rossberg-chromium** [comment]
Per-assertion IDs may not be too relevant for the core suite, since that's split into many files. In practice, the file names probably provide stable IDs with sufficient granularity (or we can make them so).

(I'd actually think it would be advantageous if the JS tests was split up into smaller pieces as well.)

**Comment by eholk** [comment]
Ah, I like the idea of using filenames. We should move the JS API tests in
that direction too, and I think that will give a reasonable stable ID for
my purposes.

**Comment by rossberg-chromium** [comment]
@eholk, is this PR dead?

**Comment by eholk** [comment]
I still think we should have a way of uniquely identifying name, but this doesn't seem like the most ideal way to do it. Maybe I'll revisit it if and when we get agreement on how to do that.

**Pull request closed without merging**